### PR TITLE
feat: verify Lean toolchain before build

### PR DESCRIPTION
**Pull request by tydeu**
---

Add an extra Lake target to ensure that Mathlib is not built on an incorrect toolchain. 

The check can be disabled with a configuration option for debugging purposes (and with the new lakefile olean cache, the option will apply to VSCode as well).

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by digama0** [comment]
I think this should not be a target, it should be a feature in `lake exe cache` instead, because the main issue caused by mismatching toolchains is the cache being useless.

**Comment by tydeu** [comment]
@digama0 The problem is that the VSCode extension automatically runs a build so, without this, set the wrong toolchain and open a file will trigger a mathlib build. With this, it will just error.

**Comment by digama0** [comment]
That's a separate issue. You also get a build triggered when you modify a file deep in the hierarchy, or touch the lakefile. I would rather we address the issues with vscode auto-build in a separate issue.

File: lakefile.lean [hunk 1]
```diff
@@ -50,6 +51,20 @@ require aesop from git "https://github.com/JLimperg/aesop" @ "master"
...
+
+/-- Verify that the workspace's toolchain matches Mathlib's. If not, replace it and error. -/
+target checkToolchain pkg : PUnit := do
+  unless noToolchainCheck do
+    let wsToolchainFile := (← getRootPackage).dir / "lean-toolchain"
+    let mathlibToolchain ← IO.FS.readFile <| pkg.dir / "lean-toolchain"
+    unless (← IO.FS.readFile wsToolchainFile).trim = mathlibToolchain.trim do
+      IO.FS.writeFile wsToolchainFile mathlibToolchain
+      error <|
+        "the workspace's toolchain does match mathlib's, so it has been replaced; " ++
```
Comment (utensil) [inline lakefile.lean:64]:
```suggestion
        "the workspace's toolchain doesn't match mathlib's, so it has been replaced; " ++
```

File: lakefile.lean [hunk 2]
```diff
@@ -50,6 +51,20 @@ require aesop from git "https://github.com/JLimperg/aesop" @ "master"
...
+        "you will need to rerun this `lake` command  to use the new toolchain"
```
Comment (utensil) [inline lakefile.lean:65]:
```suggestion
        "you will need to rerun this `lake` command  to use the new toolchain;" ++
        "remember to commit the updated toolchain file to git to ensure the consistency."
```

**Comment by tydeu** [comment]
From @digama0's comments, it seems like #7652 and leanprover/lean4#2651 + leanprover/vscode-lean4#333 have/will solve this issue in the way the mathlib maintainers want. So, I am closing this PR.

**Pull request closed without merging**